### PR TITLE
Added function docstrings; refactored a bit

### PR DESCRIPTION
**Pull request by tymorrow**
1. Simple docstrings were added to the functions in `assigner.py`, which I stole from the README.
2. I refactored the functions in assigner.py to have variable declarations at the topmost level that made sense.  The reason was that `args` alone is not a good indicator of what is available.  Having the variable declarations at the top at least gives you an idea of what you're working with.  However, I still think it's bad; I'd like to refactor this even more in the future to be more explicit.
3. Tracebacks now show up.  All errors were getting handled just fine in main, but no tracebacks were being output.  What I did seems to work just fine, but some feedback on this would be nice.
4. Converted the older % string formatting to `.format()` for consistency.
5. Implemented dry-run for `assign` subcommand.
6. Implemented verbose mode for `assign` (the way it worked before) and a progress bar as the non-verbose mode.

### CODE REVIEW

File: assigner.py [hunk 1]
```diff
@@ -20,173 +22,257 @@
...
+        base = BaseRepo(host, namespace, hw_name, token)
+        if not dry_run:
+            base.clone_to(tmpdirname, branch)
+
+        actual_count = 0  # Represents the number of repos actually pushed to
+        s_count = len(roster)
+        print("Assigning {} to {} student{} in {}...".format(
+            hw_name, s_count,
+            "s" if s_count != 1 else "",
+            "section " + section if section else "any section")
```
Comment (LinuxMercedes) [inline assigner.py]: s/any section/all sections/ if I understand correctly?

File: assigner.py [hunk 2]
```diff
@@ -20,173 +22,257 @@
...
+        )
+        for i, student in enumerate(roster):
+            username = student['username']
+            s_section = student['section']
```
Comment (LinuxMercedes) [inline assigner.py]: I would probably name this `student_section`

File: assigner.py [hunk 3]
```diff
@@ -20,173 +22,257 @@
...
     with config(args.config) as conf, tempfile.TemporaryDirectory() as tmpdirname:
-        base = BaseRepo(conf['gitlab-host'], conf['namespace'], args.name, conf['token'])
-        if args.branch:
-            base.clone_to(tmpdirname, args.branch)
+        host = conf['gitlab-host']
+        namespace = conf['namespace']
+        token = conf['token']
+        semester = conf['semester']
+        if section:
+            roster = [s for s in conf['roster'] if s['section'] != section]
```
Comment (LinuxMercedes) [inline assigner.py]: Thanks for un-hacking this =D

File: assigner.py [hunk 4]
```diff
@@ -20,173 +22,257 @@
...
+            "s" if s_count != 1 else "",
+            "section " + section if section else "any section")
+        )
+        for i, student in enumerate(roster):
+            username = student['username']
+            s_section = student['section']
+            full_name = StudentRepo.name(semester, s_section, hw_name, username)
+            repo = StudentRepo(host, namespace, full_name, token)
+
+            # Is it always necessary to log this?  Feels wrong.
```
Comment (LinuxMercedes) [inline assigner.py]: The key bit here is the line `repo.info`. This try-catch attempts to get an already existing repo; if it doesn't exist, `repo.info` will throw and the exception will handle creating the repo. If we get past `repo.info`, then the repo must already exist. 

If you couldn't tell, this feature was added an hour before I needed it for a lab.

File: assigner.py [hunk 5]
```diff
@@ -20,173 +22,257 @@
...
-                logging.warning("Student repository %s already exists.", repo.name)
-
-                if args.force:
-                    logging.warning("Deleting...")
-                    repo.delete()
-                    repo = StudentRepo.new(base, conf['semester'], student['section'], student['username'], conf['token'])
-
-                    if args.branch:
+                if force:
+                    logging.warning("Deleting...") if verbose else None
```
Comment (LinuxMercedes) [inline assigner.py]: I think something as important as deleting a student repo should result in output...assigner isn't trying to be `rm`, so we can make a little noise!

File: assigner.py [hunk 6]
```diff
@@ -20,173 +22,257 @@
...
-                    if args.branch:
-                        repo.push(base, args.branch)
-                        count += 1
+                    if branch:
+                        if not dry_run:
+                            repo.push(base, args.branch)
+                        actual_count += 1
                     else:
-                        logging.warning("Skipping...")
+                        logging.warning("Skipping...") if verbose else None
```
Comment (LinuxMercedes) [inline assigner.py]: Same comment as above.

File: assigner.py [hunk 7]
```diff
@@ -20,173 +22,257 @@
...
-                    repo.delete()
-                    repo = StudentRepo.new(base, conf['semester'], student['section'], student['username'], conf['token'])
-
-                    if args.branch:
+                if force:
+                    logging.warning("Deleting...") if verbose else None
+                    if not dry_run:
+                        repo.delete()
+                        repo = StudentRepo.new(base, semester, s_section, username, token)
                         repo.push(base, args.branch)
```
Comment (LinuxMercedes) [inline assigner.py]: `args.branch` -> `branch`?

File: assigner.py [hunk 8]
```diff
@@ -20,173 +22,257 @@
...
+        semester = conf['semester']
+
         count = 0
-        for student in conf['roster']:
-            if args.section and student['section'] != args.section:
+        for student in roster:
+            username = student['username']
+            s_section = student['section']
+
+            if section and s_section != section:
```
Comment (LinuxMercedes) [inline assigner.py]: This should be un-hacked similar to the list expression above. Also, same comment wrt `s_section`.

File: assigner.py [hunk 9]
```diff
@@ -20,173 +22,257 @@
...
         os.makedirs(path, mode=0o700, exist_ok=True)
 
         count = 0
-        for student in conf['roster']:
-            if args.section and student['section'] != args.section:
+        for student in roster:
+            username = student['username']
+            s_section = student['section']
+
+            if section and s_section != section:
```
Comment (LinuxMercedes) [inline assigner.py]: Again with un-hacking...I should probably do that myself since it's my fault in the first place.

File: assigner.py [hunk 10]
```diff
@@ -20,173 +22,257 @@
...
-        for student in conf['roster']:
-            if args.section and student['section'] != args.section:
-                continue
+        for student in roster:
+            username = student['username']
+            s_section = student['section']
 
-            name = StudentRepo.name(conf['semester'], student['section'], args.name, student['username'])
+            if section and s_section != section:
+                continue
```
Comment (LinuxMercedes) [inline assigner.py]: hax again

File: assigner.py [hunk 11]
```diff
@@ -401,9 +502,10 @@ def main():
     try:
         args.run(args)
     except Exception as e:
-        if args.tracebacks:
-            raise e
-        logger.error(str(e))
+        logger.error(e)
```
Comment (LinuxMercedes) [inline assigner.py]: Shouldn't this be wrapped in `if args.tracebacks:`?

**Comment by LinuxMercedes** [comment]
Looks good! Thanks for cleaning up a lot of my mess. 

I'm not sure what to think about having a `--verbose` flag for assign; a lot of the stuff assigner does is not reversible (and idempotence only gets you so far), so re-running assigner won't always get you the same results. So you can't run assigner, see something odd that you need `--verbose` on to figure out, and then re-run with the same results. I'd be okay with it as a config option, but I would prefer to err on the side of too much data with a summary at the end rather than too little. 

For tracebacks, you need the `--tracebacks` flag; the ordering is weird since it's a flag for assigner overall, not a subcommand, so you have to call it something like `python assigner.py --tracebacks assign lab04`. It'd be nice to eventually fix the ordering problem.

**Comment by LinuxMercedes** [comment]
This seems to be working for me so far.

File: baserepo.py [hunk 12]
```diff
@@ -53,52 +56,53 @@ def from_url(cls, url, token):
...
         namespace = match.group("namespace")
         name = match.group("name")
 
-        self = cls(urlunsplit((parts.scheme, parts.netloc, '', '', '')), namespace, name, token, url)
+        self = cls(urlunsplit((parts.scheme, parts.netloc, "", "", "")),
+                   namespace, name, token, url)
 
         logging.debug(json.dumps(self.info))
-        logging.info("Found %s", self.name_with_namespace)
+        logging.info("{} is valid.".format(self.name_with_namespace))
```
Comment (LinuxMercedes) [inline baserepo.py]: Should this be a debug-level log? It is weird seeing it in the output all the time.

Comment (tymorrow) [inline baserepo.py]: Sure, makes sense to me.

**Comment by LinuxMercedes** [comment]
LGTM :+1:

**Comment by LinuxMercedes** [comment]
Do you want to clean up the commit history a bit? If not, we can probably merge this.

**Comment by tymorrow** [comment]
You mean squash it?

**Comment by LinuxMercedes** [comment]
You've done quite a lot of work here, so I wouldn't squash it into one commit, but it might be wise to rearrange and squash some commits together. You can do that with an interactive rebase pretty easily.

**Comment by tymorrow** [comment]
Commits cleaned up!

**Pull request merged**